artemis-commons/src/main/java/org/apache/activemq/artemis/utils/collections/PriorityCollection.java

```diff
@@ -0,0 +1,319 @@
+      }
+      return iterators;
+   }
+
+   @SuppressWarnings("unchecked")
+   private static <T> ResettableIterator<T>[] newResettableIteratorArrayInstance(int length) {
+      return (ResettableIterator<T>[]) Array.newInstance(ResettableIterator.class, length);
+   }
+
+   @Override
+   public void forEach(Consumer<? super T> action) {
+      Objects.requireNonNull(action);
+      PriorityHolder<T>[] current = getArray();
+      int len = current.length;
+      for (int i = 0; i < len; ++i) {
+         current[i].getValues().forEach(action);
+      }
+   }
+
+   private Collection<T> getCollection(int priority, boolean createIfMissing) {
+      PriorityHolder<T>[] current = getArray();
+      int low = 0;
+      int high = current.length - 1;
+
+      while (low <= high) {
+         int mid = (low + high) >>> 1;
+         PriorityHolder<T> midVal = current[mid];
+
+         if (midVal.getPriority() > priority)
+            low = mid + 1;
+         else if (midVal.getPriority() < priority)
+            high = mid - 1;
+         else
+            return midVal.getValues(); //key found
+      }
+
+      if (createIfMissing) {
+         PriorityHolder<T>[] newArray = newPrioritySetArrayInstance(current.length + 1);
+         if (low > 0) {
+            System.arraycopy(current, 0, newArray, 0, low);
+         }
+         if (current.length - low > 0) {
+            System.arraycopy(current, low, newArray, low + 1, current.length - low);
+         }
+         newArray[low] = new PriorityHolder<T>(priority, supplier);
+         setArray(newArray);
+         return newArray[low].getValues();
+      }
+      return null;
+   }
+
+   @Override
+   public synchronized boolean add(T t) {
+      boolean result = addInternal(t);
+      calcSize();
+      return result;
+   }
+
+   private boolean addInternal(T t) {
+      if (t == null) return false;
+      Collection<T> priority = getCollection(t.getPriority(), true);
+      return priority.add(t);
+   }
+
+   @Override
+   public boolean remove(Object o) {
+      return o instanceof PriorityAware && remove((PriorityAware) o);
+   }
+
+   public synchronized boolean remove(PriorityAware priorityAware) {
+      boolean result = removeInternal(priorityAware);
+      calcSize();
+      return result;
+   }
+
+   private boolean removeInternal(PriorityAware priorityAware) {
+      if ( priorityAware == null) return false;
+      Collection<T> priority = getCollection(priorityAware.getPriority(), false);
+      boolean result = priority != null && priority.remove(priorityAware);
+      if (priority != null && priority.size() == 0) {
+         removeCollection(priorityAware.getPriority());
+      }
+      return result;
+   }
+
+   private Collection<T> removeCollection(int priority) {
+      PriorityHolder<T>[] current = getArray();
+      int len = current.length;
+      int low = 0;
+      int high = len - 1;
+
+      while (low <= high) {
+         int mid = (low + high) >>> 1;
```

Comment:
This is just coding pref for this stuff im happier with shift operations. Its how alot of the jvm classes do this.

Reply:
I know but the reason is because some optimisations were not present in old JDK version and the classes aren't changed...

Reply:
still just a coding preference, i see you even do bitwise alot when can be just written in standard math ;) ..... 

anyhow have changed to avoid disagreement.